<!DOCTYPE html>
<html lang="en">
<head>
	<title>Genomics in R by examples</title>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="lib/highlight/styles/github.css" type="text/css"/>
	<!--
	<link rel="stylesheet" href="lib/katex/katex.min.css" type="text/css"/>
	<link rel="stylesheet" href="lib/katex/contrib/copy-tex.min.css" type="text/css"/>
	-->
	<link rel="stylesheet" href="styles/styles.css">

	<!-- Twitter Card -->
	<!-- Ref: https://dev.twitter.com/cards/types/summary-large-image -->
	<meta name="twitter:card" content="summary_large_image">
	<meta name="twitter:site" content="@ccwang002">
	<meta name="twitter:creator" content="@ccwang002">
	<meta name="twitter:title" content="Genomics in R by Examples">
	<meta name="twitter:description" content="Showcase the R/Bioconductor ecosystem for genomic analysis">
	<meta name="twitter:image" content="https://blog.liang2.tw/2019Talk-Genomics-in-R/pictures/cover.jpg">
</head>
<body class="shower list">

	<header class="caption">
		<h1>Genomics in R by Examples</h1>
		<p><a href="https://blog.liang2.tw">Liang-Bo Wang</a>, 2019-03-27.</p>
	</header>

	<!-- Cover Slide -->
	<section id="cover" class="slide clear black">
		<h2 id="talk-header">Genomics in R by Examples</h2>
		<p id="talk-author">
			<a href="https://blog.liang2.tw">Liang-Bo Wang</a><br>
			Shared under CC 4.0 BY license
		</p>
		<p id="usage-instr">
			<kbd>Esc</kbd> to overview <br />
			<kbd>←</kbd> <kbd>→</kbd> to navigate
		</p>
		<img src="pictures/cover.jpg" class="cover w" alt="">
		<style>
			#talk-header {
				color: #FFFFFF;
				font-size: 4rem;
				text-shadow: 1px 1px 3px #000000;
				text-align: center;
				left: 3rem;
				right: 3rem;
				top: 150px;
				position: absolute;
			}
			#talk-author {
				color: white;
				font-size: 2rem;
				line-height: 2em;
				left: 4rem;
				bottom: 0.5rem;
				position: absolute;
			}
			#talk-author a {
				color: #FCB100;
			}
			#usage-instr {
				position: absolute;
				text-align: right;
				font-size: 2rem;
				line-height: 2em;
				color: white;
				right: 4rem;
				bottom: 0.5rem;
			}
			#usage-instr kbd {
				color: #606060;
				background-color: white;
			}
		</style>
	</section>

	<!-- TOC
	- Examples:
		- Conversion from ENST/ENSG/ENSP to the gene symbol
			- TxDb; ensembldb
		- Extract the genomic location of a given gene
			- GRanges/GRangesList
			- Seqinfo
		- Find coverage (sequencing depth) of a gene
			- GRanages operation
			- gtrellis
		- Gene expression / CNV of the whole cohort
			- SummarizedExperiment
			- Using airway dataset as demo
			- TCGAbiolinks for all harmonized and legacy TCGA data
	- R ecosystem for genomic analysis
	Example DNA WXS BAM
	- chr22: CHEK2 SOX10 EP300 XRCC6 BCR
	-->

	<section id="goal" class="slide">
		<h2>R has a powerful genomic toolbox</h2>
		<ul>
			<li>Query annotations like genome sequence, transcript, and array probes</li>
			<li>Process genomic ranges (from annotations)</li>
			<li>Store experiment results with annotations and metadata</li>
		</ul>
		<p>Plus all the R packages built on the strong foundation:</p>
		<div class="columns three">
			<p>Visualization</p>
			<p>Statistical analysis</p>
			<p>Dataset import</p>
		</div>
	</section>

	<section class="slide">
		<h2>Get a taste by examples</h2>
		<ol>
			<li>Conversion between gene/transcript IDs and symbols</li>
			<li>Extract and process the genomic location of a gene</li>
			<li>Plot whole genome/exome sequencing coverage of a gene</li>
			<li><del>Query TCGA omics using <code>TCGAbiolinks</code></del></li>
		</ol>
		<p>Check out <a href="#additional-readings">additional readings</a> for a thorough introduction</p>
	</section>

	<section id="r-setup" class="slide tight">
		<h2>R setup</h2>
		<pre class="lang-r">
<code>install.packages(c(
	"tidyverse",   # dplyr, ggplot2, tibble, stringr, ...
	"BiocManager"  # Bioconductor installer
))
BiocManger::install(c(
	"ensembldb", "EnsDb.Hsapiens.v86",
	"gtrellis", "TCGAbiolinks"
))</code>
		</pre>
	</section>

	<!-- Example: Find the gene symbol of a txid -->
	<section class="slide">
		<h2>Example: Find the gene symbol of a transcript</h2>
		<p><mark>EnsDb</mark> stores the Ensembl annotation of a specific release of an organism.</p>
		<pre class="lang-r">
<code>library(ensembldb)
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86</code>
		</pre>
	</section>

	<section class="slide tight">
		<p><code>ensembldb::select()</code> maps the IDs between columns (keytypes). </p>
		<div class="columns two l60">
			<pre class="lang-r">
<code>ensembldb::select(
    edb,
    keys = c("ENST00000396884"),
    keytype = "TXID",
    columns = c("SYMBOL")
)
#              TXID SYMBOL
# 1 ENST00000396884  SOX10</code>
			</pre>
			<pre class="lang-r"><code>keytypes(edb)
# ENTREZID
# PROTEINID
# SEQNAME
# SYMBOL
# TXBIOTYPE
# TXID
# ...</code></pre>
		</div>
	</section>

	<section id="annotaiton-source" class="slide tight">
		<h2>Annotation sources in Bioconductor</h2>
		<figure>
		<table class="striped">
		<thead>
			<tr>
				<th>Object type</th>
				<th>Example packages</th>
			</tr>
		</thead>
		<tbody>
		<tr>
			<th role="row">TxDb</th>
			<td>
				<a href="https://bioconductor.org/packages/EnsDb.Hsapiens.v86/">EnsDb.Hsapiens.v86</a>,<br>
				<a href="http://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene/">TxDb.Hsapiens.UCSC.hg38.knownGene</a>
			</td>
		</tr>
		<tr>
			<th role="row">OrgDb</th>
			<td>
				<a href="http://bioconductor.org/packages/org.Hs.eg.db/">org.Hs.eg.db</a>
			</td>
		</tr>
		<tr>
			<th role="row">BSgenome</th>
			<td>
				<a href="http://bioconductor.org/packages/BSgenome.Hsapiens.UCSC.hg38/">BSgenome.Hsapiens.UCSC.hg38</a>
			</td>
		</tr>
		<tr>
			<th role="row">Misc.</th>
			<td>
				<a href="http://bioconductor.org/packages/Organism.dplyr/">Organism.dplyr</a>,
				<a href="http://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a>,
				<a href="http://bioconductor.org/packages/biomaRt/">biomaRt</a>,<br>
				all packages under <a href="https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData">AnnotationData</a>
			</td>
		</tr>
		</tbody>
		</table>
			<figcaption class="citation right">
				Ref: Bioconductor annotation workflow&mdash;<a href="https://bioconductor.org/packages/release/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">Genomic Annotation Resources</a>.
			</figcaption>
		</figure>
		</div>
	</section>

	<section class="slide">
		<h2>What about RefSeq transcript ID?</h2>
		<p>Try another annotation source like <a href="https://bioconductor.org/packages/org.Hs.eg.db/"><code>org.Hs.eg.db</code></a>.</p>
		<p><a href="https://bioconductor.org/packages/biomaRt/"><code>biomaRt</code></a> can convert IDs between different annotation sources.</p>
		<p>Use <code>GenomicFeatures::makeTxDbFromGFF</code> to build a new TxDB
			from <a href="ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.gff.gz">RefSeq's GFF file</a>.
			This method applies to all customized GTFs.
		</p>
	</section>

	<!-- Example: Find the transcript location -->
	<section class="slide tight">
		<h2>Example: Locate my transcript</h2>
		<pre class="lang-r small">
			<code>txs <- <mark>transcripts</mark>(edb, columns = c('symbol')); txs
# GRanges object with 216741 ranges and 1 metadata columns:
#                   seqnames            ranges strand |      symbol
#                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#   ENST00000456328        1       11869-14409      + |     DDX11L1
#   ENST00000450305        1       12010-13670      + |     DDX11L1
#   ENST00000488147        1       14404-29570      - |      WASH7P
#               ...      ...               ...    ... .         ...
#   ENST00000435945        Y 26594851-26634652      - |     PARP4P1
#   ENST00000435741        Y 26626520-26627159      - |     FAM58CP
#   ENST00000431853        Y 56855244-56855488      + |     CTBP2P1
#   -------
#   seqinfo: 357 sequences from GRCh38 genome</code>
		</pre>
		<p class="next">How to read this output? What is a <code>GRanges</code> object?</p>
	</section>

	<section id="granges-overview" class="slide">
		<h2><code>GRanges</code> overview</h2>
		<ul>
			<li><code>GRanges</code> and <code>GRangesList</code> store genomic ranges in R</li>
			<li>Provided by <a href="http://bioconductor.org/packages/GenomicRanges/">GenomicRanges</a> package</li>
			<li>Coordinate system:
			<ul>
				<li><strong>1-based</strong>: First base of the chromosome starts with coordinate 1</li>
				<li><strong>Closed</strong>: Start and end coordinates are included in the range</li>
				<li><strong>Left-most</strong>: Start &le; end regardless of the strand</li>
			</ul>
			</li>
		</ul>
	</section>

	<section id="granges-cheatsheet" class="slide tight">
		<h2><code>GRanges</code> cheatsheet</h2>
		<figure>
			<img class="fill w" src="pictures/GRanges_cheatsheet.png" alt="">
			<figcaption class="citation right">Ref: <a href="http://bioconductor.org/help/course-materials/2015/Uruguay2015/V3-GenomicRanges.html">Genomic Ranges For Genome-Scale Data And Annotation</a>, Martin Morgan, 2015.10</p>
		</div>
	</section>

	<section id="granges-example" class="slide tight">
		<p>Find the location of all SOX10 transcripts:</p>
		<pre class="lang-r small"><code>txs[txs$symbol == 'SOX10', ]
# GRanges object with 6 ranges and 1 metadata columns:
#                   seqnames            ranges strand |      symbol
#                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#   ENST00000446929       22 37970686-37983414      - |       SOX10
#   ENST00000396884       22 37972300-37984537      - |       SOX10
# ...</code></pre>
		<p>Calculate their promoter region:</p>
		<pre class="lang-r small"><code>promoters(subset(txs, symbol == 'SOX10'), upstream=2000, downstream=200)
# GRanges object with 6 ranges and 1 metadata columns:
#                   seqnames            ranges strand |      symbol
#                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
#   ENST00000446929       22 37983215-37985414      - |       SOX10
#   ENST00000396884       22 37984338-37986537      - |       SOX10
# ...
</code></pre>
		<style>
			#granges-example p {
				margin-bottom: 0.25em;
			}
		</style>
	</section>

	<section class="slide">
		<h2>GRanges/GRangesList use cases</h2>
		<ul>
			<li>Gene model (exons of a transcript)</li>
			<li>Location of microarray probes and other annotations</li>
			<li>Store experiment results:<br>
				sequencing depth, copy number change, ChIP-seq peaks</li>
		</ul>
	</section>

	<!-- Example: Plot sequencing depth -->
	<section class="slide">
		<h2>Example: Plot the whole exome sequencing depth around SOX10</h2>
		<pre class="lang-r"><code>library(rtracklayer)
library(GenomicRanges)
library(gtrellis)
library(EnsDb.Hsapiens.v86)
library(tidyverse)</code></pre>
	</section>

	<section class="slide">
		<pre class="lang-r small"><code># Read sequencing depth BED file as a GRanges object
seq_depth_gr <- import.bedGraph('wxs_normal.subset.bed.gz')
# GRanges object with 76574 ranges and 1 metadata column:
#           seqnames            ranges strand |     score
#       [1]    chr22 23174700-23174716      * |         1
#       [2]    chr22 23174717-23174725      * |         2
#       [3]    chr22 23174726-23174735      * |         3
# ...

# Find the genomic range of SOX10
edb <- EnsDb.Hsapiens.v86
seqlevelsStyle(edb) <- 'UCSC'
sox10_region <- range(transcripts(edb, filter = ~ symbol == 'SOX10')) + 500
# GRanges object with 1 range and 0 metadata columns:
#       seqnames            ranges strand
#   [1]    chr22 37970186-37987922      -
</code></pre>
	</section>

	<section class="slide tight">
		<p>Finally, use <a href="http://bioconductor.org/packages/gtrellis/">gtrellis</a> to plot. gtrellis visualizes tracks of data using chromosomes as x-axis.
			Useful for CNV, sequencing depths, DNA methylation, and etc.</p>
		<pre class="lang-r small"><code>gtrellis_layout(
    data = sox10_region,  # specify the genomic region of interest
    track_ylim = c(0, 750)
)

# Add the track for WXS sequencing depth
add_lines_track(
    seq_depth_gr,
    seq_depth_gr$score,
    area = TRUE,
    gp = gpar(fill = "gray", col = NA)
)</code></pre>
	</section>

	<section class="slide tight">
		<img class="fill w" src="pictures/seq_depth.png" alt="">
	</section>

	<!-- SummarizedExperiment -->
	<section id="summarizedexperiment-overview" class="slide tight">
		<h2>Store -omics data with its metadata</h2>
		<div class="columns two">
			<div>
				<ul>
					<li>Keep measurements of all samples with annotations, sample info, and metadata</li>
					<li>Use cases: gene expression, microarray data</li>
					<li><a href="https://bioconductor.org/packages/SummarizedExperiment/">SummarizedExperiment</a> provides <code>SummarizedExperiment</code> and <code>RangedSummarizedExperiment</code></li>
				</ul>
			</div>
			<figure>
				<img class="fill w" src="pictures/summarizedexperiment_overview.svg" alt="">
				<figcaption class="citation right">
					Ref: <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment's vignette</a>
				</figcaption>
			</figure>
		</div>
		<style>
			#summarizedexperiment-overview .columns {
				grid-column-gap: 2em;
			}
		</style>
	</section>

	<section id="summary" class="slide">
		<h2>Summary</h2>
		<ul>
			<li>Query various annotations by R using <code>TxDb</code> and <code>EnsDB</code></li>
			<li>Store genomic ranges by <code>GRanges</code> and <code>GRangesList</code></li>
			<li>Plot data tracks along genome by <code>gtrellis</code></li>
			<li>Store multi-sample experiments by <code>SummarizedExperiment</code></li>
		</ul>
	</section>

	<!-- Example: TCGAbiolinks to query gene expression of GBM samples -->
	<!--
	<section class="slide">
	  <pre class="lang-r">
<code>barcodes <- primary_tumors_tbl$barcode
meth_status_tbl <- primary_tumors_tbl %>%
    select(barcode,
	       MGMT_promoter = subtype_MGMT.promoter.status)
# Subset the gene expression assay
mgmt_exp_tbl <- assay(gene_exp_se[mgmt_gene_id, barcodes]) %>%
    as_tibble() %>%
    gather(barcode, fpkm_uq) %>%
    left_join(meth_status_tbl, by = 'barcode')</code>
		</pre>
	</section>

	<section id="gbm-mgmt-fig-code" class="slide tight">
		<h2>Plot by ggplot2</h2>
		<pre class="lang-r">
<code>ggplot(mgmt_exp_tbl,
       aes(x = MGMT_promoter, y = fpkm_uq,
           fill = MGMT_promoter)) +
    scale_y_log10(labels = scales::comma) +
    geom_violin(draw_quantiles = .50) +
    geom_quasirandom() +  # Require ggbeeswarm package
	theme_bw()</code>
		</pre>
	</section>
	-->

	<section id="tcgabiolinks" class="slide tight">
		<h2>Extra example: Access TCGA data by <a href="http://bioconductor.org/packages/TCGAbiolinks/">TCGAbiolinks</a></h2>
		<p>Visualize the relationship between DNA methylation of MGMT promoter and its gene expression.</p>
		<ul>
			<li>Query the gene expression of TCGA GBM primary tumors and store it as a <code>SummarizedExperiment</code> object</li>
			<li>Use the sample information in <code>colData()</code> to stratify GBM samples by their MGMT methylation status</li>
		</ul>
	</section>

	<section id="gbm-mgmt-fig" class="slide">
		<img class="place c" src="pictures/mgmt_promoter.png" alt="">
		<style>
			#gbm-mgmt-fig img {
				height: 95%;
			}
		</style>
	</section>

	<!-- Ending Slide -->
	<section id='end' class='slide clear black'>
		<h2>Thank you</h2>
		<img src="pictures/cover.jpg" class="cover w" alt="">
		<style>
			#end h2 {
				color: #FFFFFF;
				font-size: 4rem;
				text-shadow: 1px 1px 3px #000000;
				text-align: center;
				left: 3rem;
				right: 3rem;
				top: 150px;
				position: absolute;
			}
		</style>
	</section>

	<!-- Section: Additional readings -->
	<section id="additional-readings" class="slide">
		<h2>Additional readings</h2>
		<ul>
			<li><a href="https://www.bioconductor.org/packages/release/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">Bioconductor Annotation Workflow</a> covers AnnotationHub, OrgDb, TxDb, BSgenome and biomaRt</li>
			<li><a href="https://blog.liang2.tw/posts/2015/12/biocondutor-genomic-data/">My post</a> and <a href="https://blog.liang2.tw/2016Talk-Genomics-in-R/#">my slides</a> about the Bioconductor ecosystem in general</li>
			<li><a href="https://blog.liang2.tw/posts/2016/05/biocondutor-ensembl-reference/">My another post</a> focusing on Ensembl annotations</li>
			<li><a href="https://genomicsclass.github.io/book/">PH525x series - Biomedical Data Science</a></li>
			<li><a href="https://kasperdanielhansen.github.io/genbioconductor/">Bioconductor for Genomic Data Science</a></li>
		</ul>
	</section>

	<!-- About me -->
	<section id="about-me" class="slide clear">
		<div class="columns two">
			<div>
				<h2>About Me</h2>
				<ul>
					<li>Liang-Bo Wang</li>
					<li>Computational and Systems Biology PhD student at Washington University in St. Louis (WashU; WUSTL)</li>
					<li>Advisor: <a href="https://dinglab.wustl.edu/">Li Ding</a></li>
					<li>Research interest: cancer genomics, proteomics, single cell imaging</li>
				</ul>
			</div>
			<div>
				<img class="fill w" src="pictures/me.jpg" alt="My Protrait">
			</div>
		</div>
		<style>
			#about-me .columns {
				grid-template-columns: 3fr 2fr;
				grid-column-gap: 2em;
			}
		</style>
	</section>

	<div class="progress"></div>

	<!-- GitHub badge -->
	<footer class="badge">
		<a href="https://github.com/ccwang002/2019Talk-Genomics-in-R">
			<svg width="81" height="79" viewBox="0 0 81 79" aria-hidden="true">
				<path d="M40.5 0a40.5 40.5 0 0 0-12.8 78.93c2 .37 2.76-.88 2.76-2s0-3.51-.05-6.89c-11.27 2.45-13.64-5.43-13.64-5.43-1.84-4.68-4.5-5.92-4.5-5.92-3.68-2.51.28-2.46.28-2.46 4.06.29 6.2 4.17 6.2 4.17 3.61 6.19 9.48 4.4 11.79 3.37a8.65 8.65 0 0 1 2.57-5.41c-9-1-18.45-4.5-18.45-20a15.65 15.65 0 0 1 4.17-10.87 14.57 14.57 0 0 1 .4-10.72s3.4-1.09 11.14 4.15a38.39 38.39 0 0 1 20.28 0c7.73-5.24 11.13-4.15 11.13-4.15a14.55 14.55 0 0 1 .4 10.72 15.63 15.63 0 0 1 4.16 10.87c0 15.56-9.47 19-18.49 20 1.45 1.25 2.75 3.72 2.75 7.5v11.11c0 1.08.73 2.34 2.78 1.95A40.51 40.51 0 0 0 40.5 0z"/>
			</svg>
			Fork me<br>on GitHub
		</a>
	</footer>

	<script src="node_modules/shower-core/shower.min.js"></script>
	<script src="lib/highlight/highlight.pack.js" type="text/javascript" charset="utf-8"></script>
	<!--
	<script defer src="lib/katex/katex.min.js"></script>
	<script defer src="lib/katex/contrib/auto-render.min.js"></script>
	<script defer src="lib/katex/contrib/copy-tex.min.js"></script>
	-->
	<script>
		document.addEventListener('DOMContentLoaded', (event) => {
			// Convert tab to be 4 space in the code examples
			hljs.configure({
				tabReplace: '    '
			});
			// Highlight only the code blocks with specified language
			document.querySelectorAll('pre[class^="lang"] code').forEach((block) => {
				hljs.highlightBlock(block);
			});

			// Load KaTeX
			// renderMathInElement(document.body);
		});
	</script>
	<!-- Copyright © 3000 Yours Truly, Famous Inc. -->
</body>
</html>
