---
title: "Plot sequencing depth"
output: html_notebook
---

```{r, message=FALSE}
library(rtracklayer)
library(gtrellis)
library(EnsDb.Hsapiens.v86)
library(tidyverse)
```

```{r}
wxs_normal_seq_depth_gr <- import.bedGraph('../data/wxs_normal.subset.bed.gz')
```

```{r}
wxs_normal_seq_depth_gr
```


```{r}
edb <- EnsDb.Hsapiens.v86
seqlevelsStyle(edb) <- 'UCSC'
sox10_region <- range(transcripts(edb, filter = ~ symbol == 'SOX10')) + 500
sox10_region
```

```{r}
sox10_txs <- exonsBy(edb, filter = ~ symbol == 'SOX10')
```

```{r, fig.width=8, fig.height=5}
png('../pictures/seq_depth.png', width = 1600, height = 1000, res = 300)
n = 3
gtrellis_layout(
    data = sox10_region,
    track_ylim = c(c(0, 750), c(0.5, n+0.5)),
    track_axis = c(TRUE< FALSE),
    n_track = 2,
    track_ylab = c("WXS coverage", "SOX10"),
    track_height = unit.c(unit(1, 'null'), unit(1.5, 'cm')),
    add_name_track = TRUE
)
add_lines_track(
    wxs_normal_seq_depth_gr, 
    wxs_normal_seq_depth_gr$score, 
    area = TRUE,
    gp = gpar(fill = "gray", col = NA)
)
add_track(panel_fun = function(gr) {
  tr = sox10_txs[1:3, ] # all transcripts for this gene
  for(i in seq_along(tr)) {
      # for each transcript
      current_tr_start = min(start(tr[[i]]))
      current_tr_end = max(end(tr[[i]]))
      grid.lines(c(current_tr_start, current_tr_end), c(n - i + 1, n - i + 1), 
          default.units = "native", gp = gpar(col = "#CCCCCC"))
      grid.rect(start(tr[[i]]), n - i + 1, width(tr[[i]]), 0.6,
          default.units = "native", just = "left", 
          gp = gpar(fill = "orange", col = "orange"))
  }
})
dev.off()
```

```{r}
gtrellis_layout(
    data = sox10_region,
    track_ylim = c(0, 750),
    add_name_track = TRUE
)
add_lines_track(
    wxs_normal_seq_depth_gr, 
    wxs_normal_seq_depth_gr$score, 
    area = TRUE,
    gp = gpar(fill = "gray", col = NA)
)
```

